### ISSUE TITLE
Allow Terminals To Be Opened Anywhere Files Can Be Opened.

### ISSUE THREAD

**Issue by jordwalke** [Microsoft/vscode#24922]
Feature Request:
I know this has been discussed at length, but I would like to bring up something more specific. I would like terminals to have every advantage that files have in terms of positioning and windowing. That would mean, that wherever you see a file opened in VSCode we could just as easily have a terminal there. At first, the feature request is merely about positioning of terminals, but eventually I imagine creating extensions that can register themselves as "handlers" for certain file types, and then when opening those files, the extensions can open in a terminal instead of text editor, but that would feel very integrated because those terminals are positioned and navigable just like regular files are. For example, imagine you have some binary file, and VSCode wouldn't know how to display anything but the random bits. Well, you could make an extension that handles that binary file type and opens in a terminal which opens it in some useful CLI inspector tool.

Also, I imagine that just allowing terminals to be open anywhere files are, might be a far simpler solution than some of the other proposed features for custom terminal positioning. This is a way to just avoid all the bikeshedding about some one-off behavior for terminal positioning, and my hope is that it's easier to implement since it can just reuse all the file windowing code. This way, when/if VSCode gets new window positioning features, terminals automatically benefit from it, too.

cc @Tyriar 

(Somewhat related feature request, but not the same thing).
https://github.com/Microsoft/vscode/issues/10546

I apologize if this exact proposal has been discussed somewhere - please feel free to point me to that.

**Comment by Tyriar** [comment]
I'm pretty against the idea of having terminals anywhere as that makes what all the terminal commands do difficult to define:

- Toggling would toggle the last active terminal?
- What happens when you "untoggle" a terminal?
- Terminal exist both in editors and the panel?
- Focus next brings to foreground random tabs?
- Terminals are managed in the panel but they exist in "editor tabs"
- etc.

The way I envision it going is having tabs https://github.com/Microsoft/vscode/issues/10546, being able to split the tabs https://github.com/Microsoft/vscode/issues/7504 and being able to move the panel around https://github.com/Microsoft/vscode/issues/2806.

**Comment by jordwalke** [comment]
You could imagine keeping the HUD style terminal as it is, and not supporting any new proposed windows management for them - as long as you provide what this current Issue is proposing.

**Comment by Tyriar** [comment]
I think tabs for the terminals would need to happen before we even start thinking about this. Right now the main blockers there are time and the UX issue of having tabs within tabs but keeping things feeling "lightweight".

**Comment by jordwalke** [comment]
I understand and I proposed this issue because it could allows you to completely punt on most of that work. Why would you need any of those other features if tabs could just make use of all existing window positioning code for text editors?

**Comment by Tyriar** [comment]
Because it needs more thought before we make a radical change like this and there are more pressing issues to get to first. It's not just as simple as throwing the terminal functionality into editor tabs, it needs to get a team consensus for the UX, consider the impacts on the current functionality (commands, settings, API), consider the impacts on future work (tabs, splits, etc.). There are also issues out that are looking into making the workbench more flexible which could resolve this as well.
